### PR TITLE
Add support for pipe syscall (PIPEFS)

### PR DESCRIPTION
**Pull request by cynecx**
Addressing https://github.com/kripken/emscripten/issues/2883.
- [ ] Add test for pipefs
- [ ] Add more tests for pipefs

### CODE REVIEW

**Comment by kripken** [comment]
An overview might help here - what does this do, in the big picture?

**Comment by cynecx** [comment]
This is needed to get boost/asio to work with emscripten. See issue https://github.com/kripken/emscripten/issues/2883.

Every call to the pipe syscall will basically create a new fifo buffer which users can write and read to. That's actually it. Asio uses this mechanism to interrupt an event loop. Yeah, I know this sound paradox as the browser is itself event-driven, this means that the user who wants to use asio has to integrate asio's event loop into the browser's one by calling the poll_one function.

**Comment by kripken** [comment]
Ok, thanks.

File: src/library_pipefs.js [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+      var writableStream = FS.createStream({
+        path: wName,
+        node: wNode,
+        flags: FS.modeStringToFlags('w'),
+        seekable: false,
+        stream_ops: PIPEFS.stream_ops
+      });
+      wNode.stream = writableStream;
+
+      return [ readableStream.fd, writableStream.fd ];
```
Comment (kripken) [inline src/library_pipefs.js]: may as well return an object here, so you don't need to use 0 and 1 to index, but can use proper names.

File: src/library_pipefs.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    },
+    stream_ops: {
+      poll: function(stream) {
+        var pipe = stream.node.pipe;
+        var currentLength = 0;
+
+        if(pipe.buckets.length > 0) {
```
Comment (kripken) [inline src/library_pipefs.js]: space after `if`

File: src/library_pipefs.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+        var toRead = Math.min(currentLength, length);
+        if(toRead == 0) {
+          return 0;
+        }
+
+        var totalRead = toRead;
+        var toRemove = 0;
+
+        for(var i = 0; i < pipe.buckets.length; i++) {
```
Comment (kripken) [inline src/library_pipefs.js]: space after `for`

File: src/library_pipefs.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+          } else {
+            var tmpSlice = currBucket.buffer.subarray(currBucket.roffset, currBucket.offset);
+            data.set(tmpSlice);
+            data = data.subarray(tmpSlice.byteLength);
+            toRead -= tmpSlice.byteLength;
+            toRemove++;
+          }
+        }
+
+        while(toRemove--) {
```
Comment (kripken) [inline src/library_pipefs.js]: also after `while`

**Comment by kripken** [comment]
What's the status of a test for this?

**Comment by cynecx** [comment]
Addressed all comments. A test will follow soon as I am currently heavily testing asio.

**Comment by kripken** [comment]
Thanks, looks good so far. Waiting on the test, as it'll help me understand this more completely.

**Comment by juj** [comment]
Will this get correctly dead code eliminated away if one doesn't call `pipe()` in a program at all? I.e. no `syscall42`? That is worth checking, since this will be somewhat rarely used functionality, so that it won't be pulled in for builds that don't need it.

**Comment by cynecx** [comment]
Sorry about the delay, I was kinda busy these days. I will try to write some basic tests now.

@juj Do you mean we should put this behind an `#ifdef USE_PIPE`?

**Comment by kripken** [comment]
I think it should be ok: `PIPEFS` is a dependency of that syscall, and our syscalls are static, so it should only get pulled in if the syscall is actually called. We should verify to be sure, but if not, that would be a separate bug in our linking.

**Comment by juj** [comment]
Yeah, no need to have `#ifdef` if DCE works correctly already without it.

**Comment by seibelj** [comment]
👍  would appreciate a merge

**Comment by atrosinenko** [comment]
Is this PR still planned for merge? It would be great to have the `pipe` syscall working in Emscripten for the purpose of events.

**Comment by kripken** [comment]
Looks like this PR is still waiting on tests - we can't merge it without that. Also it needs to be rebased.

**Comment by atrosinenko** [comment]
I am not familiar with Emscripten FS layer, but in case you consider this implementation looking good enough, then maybe I can somehow help you with unit-testing it according to the `pipe` syscall documentation?

**Comment by cynecx** [comment]
I have rebased the branch but unfortunately I am currently unable to write tests because I simply don't have the time to setup the environment as I am very busy right now with other things.

However, it would be great if someone else could write some tests for this patch. I would really appreciate this :).

**Comment by atrosinenko** [comment]
Thank you, @cynecx. How can I contribute tests? Should this PR be integrated first into some feature branch of `kripken/emscripten` repo so I can pull it and add some tests?

**Comment by kripken** [comment]
I think the simplest thing is for you to create a branch in your repo, then merge in @cynecx's branch here, then add the tests, then make a PR with all that together.

**Comment by atrosinenko** [comment]
This PR can probably be closed because it was integrated in #4935.

**Comment by juj** [comment]
Thanks! Closing.

**Pull request closed without merging**